aurelia/aurelia
# templating tests and bugfixes

# Pull Request

## 📖 Description

This PR is a first step towards a more maintenance-friendly approach to bulk testing input combinations.


There's also a small "bugfix" for an issue uncovered by the generated tests, that is the root element's children weren't being removed on the stop task. A pretty uncommon scenario that probably wouldn't have ever bothered anyone (except for me with integration testing), but easy enough to fix that I found it worth fixing. The `element.$au` property might also very well come in handy in other situations.




### 🎫 Issues

This is part of https://github.com/aurelia/aurelia/issues/89


## 👩‍💻 Reviewer Notes

The generated tests live in the `.generated.spec.ts` files (at the time of writing only `template-compiler.generated.spec.ts`)


## 📑 Test Plan

The goals of the generated test code are pretty straight-forward:
- make sure at least some tests fail on even minor code changes (generate near-redundant combinations of small variations)
- make failing tests easy to find (sequential numbering is in place)
- make failing tests easy to debug (no chained calls, no loops, one test per `it`)
- make failing tests easy to fix (change a small piece in the generator code to fix lots of tests)


## ⏭ Next Steps


At the moment the generator code in itself contains a lot of copy-paste work with small modifications. This needs to get cleaner, but I'd like to drive this effort in a pragmatic manner. It should quickly and easily produce a high test density in complex areas that have a high chance of misbehaving in more exotic setups.

I may remove some of the existing test loops when there is no question they are fully redundant.

The current tests are mostly static. An important (and somewhat tricky) next step will be to test bound data, and data that mutates.

Other good candidates for generated tests:
- Bindings
- Observers
- View, customElement, customAttribute lifecycle hooks
- Most things, really

The plan is not to replace *all* tests with generated tests, but rather the tedious ones that are hard to debug (e.g. the loops). They should represent the thorough variants of a handful of carefully hand-crafted "model" tests, so that it's always easy to branch out / add more tests without having to dig into the generator code.
